### PR TITLE
Implement Sync Handler for Block Requests By Slot Number

### PR DESCRIPTION
**Pull request by nisdas**
This will deal with #447, which will allow nodes to respond to block requests by slot number. This builds ontop of #440
- [x] Allow nodes to receive block request protobuf messages
- [x] Scan the db by slot number and send the block to the peer who requested it
- [x] Add well-defined tests

### CODE REVIEW

File: beacon-chain/blockchain/core.go [hunk 1]
```diff
@@ -558,3 +513,89 @@ func (b *BeaconChain) initCycle(cState *types.CrystallizedState, aState *types.A
...
 	return newCrystallizedState, newActiveState
 }
+
+func (b *BeaconChain) hasBlock(blockhash [32]byte) (bool, error) {
+	return b.db.Has(blockKey(blockhash))
+}
+
+// saveBlock puts the passed block into the beacon chain db.
+func (b *BeaconChain) saveBlock(block *types.Block) error {
+
```
Comment (rauljordan) [inline beacon-chain/blockchain/core.go]: whitespace rm at top and bottom of this func

File: beacon-chain/blockchain/core.go [hunk 2]
```diff
@@ -558,3 +513,89 @@ func (b *BeaconChain) initCycle(cState *types.CrystallizedState, aState *types.A
...
+// checkForBlockBySlotNumber checks the db if the canonical block for
+// this slot exists.
+func (b *BeaconChain) checkForBlockBySlotNumber(slotnumber uint64) (bool, error) {
+	return b.db.Has(canonicalBlockKey(slotnumber))
+}
+
+// getBlockBySlotNumber retrieves the canonical block which is saved in the db
+// which has the required slot number.
+func (b *BeaconChain) getBlockBySlotNumber(slotnumber uint64) (*types.Block, error) {
+
```
Comment (rauljordan) [inline beacon-chain/blockchain/core.go]: same here

File: beacon-chain/blockchain/core.go [hunk 3]
```diff
@@ -558,3 +513,89 @@ func (b *BeaconChain) initCycle(cState *types.CrystallizedState, aState *types.A
...
+// checkForBlockBySlotNumber checks the db if the canonical block for
+// this slot exists.
+func (b *BeaconChain) checkForBlockBySlotNumber(slotnumber uint64) (bool, error) {
+	return b.db.Has(canonicalBlockKey(slotnumber))
+}
+
+// getBlockBySlotNumber retrieves the canonical block which is saved in the db
+// which has the required slot number.
+func (b *BeaconChain) getBlockBySlotNumber(slotnumber uint64) (*types.Block, error) {
```
Comment (rauljordan) [inline beacon-chain/blockchain/core.go]: slotNumber*

File: beacon-chain/blockchain/schema.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	crystallizedStateLookupKey = []byte("beacon-crystallized-state")
+
+	// GenesisLookupKey tracks the genesis block.
+	genesisLookupKey = []byte("genesis")
+
+	// Data item prefixes.
+	blockPrefix = []byte("block-") // blockPrefix + blockhash -> block
+
+	canonicalPrefix = []byte("canonical-") // canonicalPrefix + num(uint64 big endian) -> blockhash
+
```
Comment (rauljordan) [inline beacon-chain/blockchain/schema.go]: rm space

File: beacon-chain/sync/service.go [hunk 5]
```diff
@@ -154,6 +162,30 @@ func (ss *Service) run() {
...
+				continue
+			}
+
+			block, err := ss.chainService.GetCanonicalBlockBySlotNumber(request.GetSlotNumber())
+			if err != nil {
+				log.Errorf("Error retrieving block from db %v", err)
+				continue
+			}
+
+			log.WithField("slotnumber", fmt.Sprintf("%d", request.GetSlotNumber())).Debug("Sending requested block to peer")
```
Comment (rauljordan) [inline beacon-chain/sync/service.go]: slotNumber*

**Comment by prestonvanloon** [comment]
I'm not sure what the diff is between #440 and this. Will review after #440 is merged and this branch is synced

**Comment by rauljordan** [comment]
This one extends sync to be able to request blocks by slot number from other nodes @prestonvanloon

**Pull request merged**